app/models/application.rb

```diff
@@ -10,4 +10,13 @@ class Application < ApplicationRecord
 
   attribute :availability_status, :string
   validates :availability_status, :inclusion => { :in => %w[available unavailable] }, :allow_nil => true
+
+  validate :source_must_be_compatible
+
+  private
+
+  def source_must_be_compatible
+  	return if application_type.supported_source_types.include?(source.source_type.name)
+  	errors.add(:source, "is not compatible with this application type")
```

Comment:
Maybe we should include the name of the source here? "`#{source.source_type.name}` is not compatible with this application type"

Reply:
That would be useful. I'll add it.